mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp

```diff
@@ -694,6 +696,107 @@ bubbleUpPackOpThroughCollapseShape(tensor::CollapseShapeOp collapseOp,
   return success();
 }
 
+/// Project dimsPos to their collapsed positions in the reassocIndices.
+///
+/// For example, given dimsPos [0, 1, 2, 4], and matching reassocIndices
+/// [[0], [1, 2], [3], [4]], it returns [0, 1, 1, 3]. Because for pos 0,
+/// the reassoc dim [0] is 0. For pos 1 and 2, the reassoc dim in pos
+/// [1, 2] is 1. And for pos 4, the reassoc dim [4] is 3.
+static SmallVector<int64_t>
+projectDimsPosIntoReassocPos(ArrayRef<int64_t> dimsPos,
+                             ArrayRef<ReassociationIndices> reassocIndices) {
+  SmallVector<int64_t> projectedPos;
+
+  // Map each dimension to the position of corresponding reassociation index.
+  for (auto pos : dimsPos) {
+    for (auto [idx, indices] : llvm::enumerate(reassocIndices)) {
+      // If the dimension is present in the current indices group, the group
+      // position within the reassociation map is the desired projected
+      // dimension position.
+      if (llvm::any_of(indices,
+                       [&](int64_t expandDim) { return expandDim == pos; })) {
+        projectedPos.push_back(idx);
+        break;
+      }
+    }
+  }
+  assert(projectedPos.size() == dimsPos.size() && "Invalid dim pos projection");
+
+  return projectedPos;
+}
+
+/// Bubble up pack op through expand shape op.
```

Comment:
Repeating the function name as a comment is not helpful.. Can you add an example that shows the IR before the transform and the IR after the transform?

Reply:
Added an example for completeness.

Although, I don't see how it adds any value either. The name itself is verbose and quite self explanatory in context of the transform. It's also not a public facing API.
The example will only get quickly outdated.

The actually behavior is captured by the test cases and preconditions by the logic itself.
But I admit the explanations within the function were a bit superficial so, I expanded the comments and added extra test case.

Reply:
People won't need to go to lit file to see what's happening. This is the point. Can you elaborate why it will only get quickly outdated? I thought the method should always handle the case.

Reply:
Lit at least gives me guarantee that examples there are verified and valid.
I won't trust a random IR in a comment to grasp the functionality. Either, I take it at face value and "don't care", or look deeper i.e., read the logic and/or look at tests.

Perhaps, my different opinion comes from the clarity of the name itself. It's not a complex transform that requires deep explanation as it could be as long as or longer than the logic itself. The transform's name is already succinct and conveys the message.

> Can you elaborate why it will only get quickly outdated?

It's a more general statement. The simple case of `expand(pack) -> pack(expand)` stays valid. But the examples in this file are really verbose. The finer details like ops' arguments or input/output shapes might stay relevant or not.

When IR syntax or semantics change, random comments like that are least likely to get updated. In fact, expand_shape has already changed with addition of `output_shape`. It doesn't matter here but still the example is already outdated.

Reply:
agree and disagree.. It is not common to me that the IR syntax would change often, based on my experience in the MLIR community. The IR examples are more like a pseudo code, and people can also update them when they are outdated. It is also common in MLIR codebase.

The original comment is mostly from the LLVM style guide: https://www.llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments

I was hoping that we could have better comments instead of repeating the function name. Adding an example is a comment improvement to me, so I made the suggestion.

> Avoid restating the information that can be inferred from the API name.

Reply:
Definitely thanks for sharing your views here.
In the end, I'd say that neither approach is ideal but now I see where you're coming from. That should help me come up with clearer docs next time 😄